library/os_subnet.py

```diff
@@ -0,0 +1,325 @@
+    )
+
+    module_kwargs = openstack_module_kwargs()
+    module = AnsibleModule(argument_spec,
+                           supports_check_mode=True,
+                           **module_kwargs)
+
+    if not HAS_SHADE:
+        module.fail_json(msg='shade is required for this module')
+
+    state = module.params['state']
+    network_name = module.params['network_name']
+    cidr = module.params['cidr']
+    ip_version = module.params['ip_version']
+    enable_dhcp = module.params['enable_dhcp']
+    subnet_name = module.params['name']
+    gateway_ip = module.params['gateway_ip']
+    dns = module.params['dns_nameservers']
+    pool_start = module.params['allocation_pool_start']
+    pool_end = module.params['allocation_pool_end']
+    host_routes = module.params['host_routes']
+    ipv6_ra_mode = module.params['ipv6_ra_mode']
+    ipv6_a_mode = module.params['ipv6_address_mode']
+
+    # Check for required parameters when state == 'present'
+    if state == 'present':
+        for p in ['network_name', 'cidr']:
+            if not module.params[p]:
+                module.fail_json(msg='%s required with present state' % p)
+
+    if pool_start and pool_end:
+        pool = [dict(start=pool_start, end=pool_end)]
+    elif pool_start or pool_end:
+        module.fail_json(msg='allocation pool requires start and end values')
+    else:
+        pool = None
+
+    try:
+        cloud = shade.openstack_cloud(**module.params)
+        subnet = cloud.get_subnet(subnet_name)
+
+        if module.check_mode:
+            module.exit_json(changed=_system_state_change(module, subnet,
+                                                          cloud))
+
+        if state == 'present':
+            if not subnet:
+                subnet = cloud.create_subnet(network_name, cidr,
+                                             ip_version=ip_version,
+                                             enable_dhcp=enable_dhcp,
+                                             subnet_name=subnet_name,
+                                             gateway_ip=gateway_ip,
+                                             dns_nameservers=dns,
+                                             allocation_pools=pool,
+                                             host_routes=host_routes,
+                                             ipv6_ra_mode=ipv6_ra_mode,
+                                             ipv6_address_mode=ipv6_a_mode)
+                changed = True
+            else:
+                if _needs_update(subnet, module, cloud):
+                    cloud.update_subnet(subnet['id'],
+                                        subnet_name=subnet_name,
+                                        enable_dhcp=enable_dhcp,
+                                        gateway_ip=gateway_ip,
+                                        dns_nameservers=dns,
+                                        allocation_pools=pool,
+                                        host_routes=host_routes)
```

Comment:
LGTM! :+1: 

This line is the only real change... It is just a direct copying of this [PR](https://github.com/blueboxgroup/ansible-modules-core/pull/2/files), but brings the whole file into Ursula instead. This also reflects and upstream change.